addon/components/utils/base-popover.js

```diff
@@ -0,0 +1,219 @@
+
+  /**
+   * Stores the trigger element for adding/removing event listeners and data attributes
+   */
+  _triggerElement = null;
+
+  /**
+   * Stores the popper element for adding/removing event listeners and data attributes
+   */
+  _popperElement = null;
+
+  /**
+   * Used as proxy to pass along the class of this component to the actual popper
+   */
+  _popperClass = '';
+
+  constructor() {
+    super();
+
+    this._popperClass = `ice-base-popover ${this.class || ''} ${this.classNames.join(' ')}`;
+
+    for (const binding of this.classNameBindings) {
+      if (binding.value) {
+        this._popperClass += ` ${binding.value()}`;
+      }
+    }
+  }
+
+  didInsertElement() {
+    this._triggerElement = this.element.parentNode;
+    this._triggerElement.setAttribute('data-popover-trigger', guidFor(this));
+
+    if (this.get('triggerEvent') === 'click') {
+      this._triggerElement.addEventListener('mousedown', this._openPopoverHandler);
+    } else {
+      this._triggerElement.addEventListener('mouseenter', this._openPopoverHandler);
+      this._triggerElement.addEventListener('mouseleave', this._closePopoverHandler);
+    }
+  }
+
+  willDestroyElement() {
+    raf.forget(this._insertFrame);
+
+    if (this.get('triggerEvent') === 'click') {
+      this._triggerElement.removeEventListener('mousedown', this._openPopoverHandler);
+    } else {
+      this._triggerElement.removeEventListener('mouseenter', this._openPopoverHandler);
+      this._triggerElement.removeEventListener('mouseleave', this._closePopoverHandler);
+    }
+
+    this._triggerElement.removeAttribute('data-popover-trigger');
+
+    document.body.removeEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Inserts the popover and sets up the body click listener
+   */
+  _insertPopover() {
+    // Schedule these separately so that the element gets fully attached by setting
+    // `isOpen` to true, and only _then_ setting `isShowing` to true, triggering the
+    // CSS transition
+    run(() => this.set('isOpen', true));
+
+    document.body.addEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Removes the popover and tears down the body click listener
+   */
+  _removePopover() {
+    run(() => this.set('isOpen', false));
+
+    document.body.removeEventListener('mouseup', this._handleBodyClick);
+  }
+
+  /**
+   * Triggers when the popover has entered the DOM and the API has been established,
+   * allowing us to add data attributes and event handlers and mark the trigger as active
+   *
+   * @param {PopperAPI} api - The API received from the underlying popper
+   */
+  @action
+  popoverOpened({ popperElement }) {
+    this._popperElement = popperElement;
+    this._popperElement.setAttribute('data-popover-content', guidFor(this));
+    this._triggerElement.classList.add('is-active');
+
+    if (this.get('triggerEvent') === 'hover') {
+      this._popperElement.addEventListener('mouseenter', this._openPopoverHandler);
```

Comment:
Why do we want to listen to `mouseenter` when the pop over is already opened?

Reply:
this is adding the listener to the popover element itself, which just entered the DOM. It allows the user to mouse over the popover area and cancel closing.